File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/line/call/CallFunctionTest.java
package com.flat.simplex.parser.logic.line.call;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.Field;
import com.flat.simplex.parser.logic.block.BlockIf;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class CallFunctionTest {

    @Test
    public void loadFunction() {
        TokenChain chain = readChain("function(a, b) {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context);
    }

    @Test
    public void loadNoArgsFunction() {
        TokenChain chain = readChain("function() {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext());
        assertNotNull(call.getBlocks());

        assertErrors(context);
    }

    @Test
    public void loadFunctionEmptyBody() {
        TokenChain chain = readChain("function() {}");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext());
        assertNotNull(call.getBlocks());


        assertErrors(context);
    }

    @Test
    public void loadFunctionDoubleComma_Fail() {
        TokenChain chain = readChain("function(a,, b) {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void loadFunctionNoBody_Fail() {
        TokenChain chain = readChain("function(a, b)");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadFunctionMissingCloser_Fail() {
        TokenChain chain = readChain("function(a, b");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.missingCloser, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadFunctionMissingBodyCloser_Fail() {
        TokenChain chain = readChain("function(a, b) {");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.missingCloser);
    }

    @Test
    public void loadFunctionEndComma_Fail() {
        TokenChain chain = readChain("function(a, b,) {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadFunctionUnexpectedToken_Fail() {
        TokenChain chain = readChain("function(a, b) a {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a", "b");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void loadFunctionRepeatedVar_Fail() {
        TokenChain chain = readChain("function(a, a) {return a + b; }");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallFunction call = new CallFunction(blockIf, chain.get(), null);
        call.load();

        assertNotNull(call.getInnerContext(), "Invalid inner context");
        assertFields(call.getInnerContext(), "a");

        assertErrors(context, Error.varRepeatedField);
    }

    private void assertFields(Context context, String... names) {
        List<String> fields = context.getAllFieldNames();
        assertArrayEquals(names, fields.toArray(), "Invalid fields name");
    }

    private BlockIf getBlock(Context context) {
        TokenChain chain = readChain("if(true);");

        return new BlockIf(context, null, chain.get(), null);

    }
}

File: SimplexApp/simplex/lang/parser/logic/line/call/CallFunction.test.js
const TokenChain = require("simplex/lang/support/TokenChain.js");
const Context = require("simplex/lang/parser/logic/Context.js");
const ContextSupport = require("simplex/lang/support/ContextSupport.js");
const Error = require("simplex/lang/parser/logic/error/Error.js");
const CallFunction = require("simplex/lang/parser/logic/line/call/CallFunction.js");
const AssertError = require("simplex/lang/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertNotNull = AssertError.assertNotNull;
const assertArrayEquals = AssertError.assertArrayEquals;

test('loadFunction', () => {
    let chain = readChain('function(a, b) {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context);
});
test('loadNoArgsFunction', () => {
    let chain = readChain('function() {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext());
    assertNotNull(call.getBlocks());
    assertErrors(context);
});
test('loadFunctionEmptyBody', () => {
    let chain = readChain('function() {}');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext());
    assertNotNull(call.getBlocks());
    assertErrors(context);
});
test('loadFunctionDoubleComma_Fail', () => {
    let chain = readChain('function(a,, b) {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.unexpectedToken);
});
test('loadFunctionNoBody_Fail', () => {
    let chain = readChain('function(a, b)');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadFunctionMissingCloser_Fail', () => {
    let chain = readChain('function(a, b');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.missingCloser, Error.unexpectedEndOfTokens);
});
test('loadFunctionMissingBodyCloser_Fail', () => {
    let chain = readChain('function(a, b) {');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.missingCloser);
});
test('loadFunctionEndComma_Fail', () => {
    let chain = readChain('function(a, b,) {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadFunctionUnexpectedToken_Fail', () => {
    let chain = readChain('function(a, b) a {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a', 'b');
    assertErrors(context, Error.unexpectedToken);
});
test('loadFunctionRepeatedVar_Fail', () => {
    let chain = readChain('function(a, a) {return a + b; }');
    let context = new Context(chain.get());
    let call = new CallFunction(context, chain.get(), null);
    call.load();
    assertNotNull(call.getInnerContext(), 'Invalid inner context');
    assertFields(call.getInnerContext(), 'a');
    assertErrors(context, Error.varRepeatedField);
});
function assertFields(context, ...names) {
    let fields = context.getAllFieldNames();
    assertArrayEquals(names, fields, 'Invalid fields name');
}
